### DISCUSSION TITLE
dinitctl start-or-restart

### DISCUSSION THREAD

**Discussion by iacore** [davmac314/dinit#316] [General]
**Is your feature request related to a problem? Please describe.**

I had the need to use `dinitctl start X || dinitctl restart X` for a few times now.

When developing an application, and that application also runs under dinit, sometimes it crashes. I need a command to start a new version of the application (in build script) no matter if it was running or not.

I also have trouble understanding why `dinitctl restart X` doesn't start a service if it has not started.

**Describe the solution you'd like**

`dinitctl start --stop-if-running X`

**Comment by davmac314** [comment]
> I had the need to use dinitctl start X || dinitctl restart X for a few times now.

Well, `dinitctl start X` doesn't return an error if the service is already running. It should be the other way around: `dinitctl restart X || dinitctl start X` - maybe that's what you meant? Another option would be `dinitctl stop X; dinitctl start X`.

Just to make sure I understand correctly: the complaint is just that you need two commands?

> I also have trouble understanding why dinitctl restart X doesn't start a service if it has not started.

Because "restart" is not the same as "stop" + "start". There was a discussion about this a little while back but I can't find it now.

`dinitctl start` marks a service as active.

`dinitctl restart` doesn't mark a service active; it restarts a running service whether it is active or not (if it's not active, the only reason it would be running is if it has a running dependent). Since a non-active service shouldn't run, it doesn't make sense for "restart" to start a not-started service.

**Reply by iacore** [reply]
> Just to make sure I understand correctly: the complaint is just that you need two commands?

Partly yes. Need quality of life stuff.

**Comment by davmac314** [comment]
The earlier discussion was here: https://github.com/davmac314/dinit/issues/300

**Comment by iacore** [comment]
I think I don't understand how dependency / active state works in dinit. I usually just do `dinitctl enable X`. 


`man dinitctl` says:

```
       list   List loaded services and their state.  Before each service, one of the following state  indicators  is
              displayed:

...

square brackets are used if the service is marked active (target state will always be started if this is the case).
```

I don't see square brackets in the output of `dinitctl list`, so I think my services are not "active"?

**Reply by mobin-2008** [reply]
> Can you move this to Discussion?
>

I Also think it's a better place for this thing.
> I think I don't understand how dependency / active state works in dinit. I usually just do `dinitctl enable X`.
> 
> `man dinitctl` says:
> 
> ```
>        list   List loaded services and their state.  Before each service, one of the following state  indicators  is
>               displayed:
> 
> ...
> 
> square brackets are used if the service is marked active (target state will always be started if this is the case).
> ```
> 
> I don't see square brackets in the output of `dinitctl list`, so I think my services are not "active"?

Dinit is designed around a simple rule: **When a service doesn't have any purpose, We bring it down.** and this is implemented by two different type of activation: Implicit and Explicit

1. Implicit activation: When service `a` needs `b` (e.g. through a `depends-on` relationship) Dinit will start service `b` and wait's for its startup process, then Dinit will start service `a` after that. In Dinit if service `a` get stopped, also the service `b` will get stopped because service `b` doesn't have any purpose except for `a` which is stopped by now. So a implicit activated service is a service which depends on something to be active and when that thing get stopped, implicit service will not be active beyond that point.

2. Explicit activation: It simply means a service doesn't depends on something else to be active. It's self-activated.

# `[+]` vs `{+}`

It's output of `dinitctl list` on my chimera setup:
```
# dinitctl list
[[+]     ] boot
[{+}     ] system
[{+}     ] login.target (has console)
[{+}     ] local.target
[{+}     ] pre-local.target
...
```
When a service is marked with `[+]` (such as boot) that means it's explicitly activated.
When a service is marked with `{+}` (such as system) that means it's implicitly activated and in this case if boot stops this will automatically stops.

See https://github.com/davmac314/dinit/blob/master/doc/getting_started.md for some examples of this.

**Reply by davmac314** [reply]
> I think I don't understand how dependency / active state works in dinit. I usually just do dinitctl enable X.

In Dinit parlance "active" and "enabled" are different things. "Enabled" means, essentially, "is a dependency of boot (or some other target that is expected to be active), and is intended to run during normal operation", whereas "active" means "is intended to be running right now". There are two types of active: explicit and implicit (via dependency).

A service that is enabled will normally be active, but normally only implicitly active. That means it will start, but it will stop automatically if every service that it depends on it stops.

If a service is instead _explicitly_ active, it stops only if it that explicit activation is removed _and_ no other service depends on it. (Explicit activation is removed if a service stops, for any reason, and is not immediately restarted. If a service is restarted, the explicit activation is not affected).

When you execute `dinit start X` it marks the service `X` as active (explicitly). That means it will start, if it was not already active.

When you execute `dinit restart X` it restarts the service `X`. This is only possible if the service is active (explicitly or implicitly) because an inactive service is not running and should not run. Restarting a service doesn't affect its explicit activation, because sometimes you need to restart a service (that happens to be running) without explicitly activating it.

**Reply by davmac314** [reply]
> I don't see square brackets in the output of dinitctl list, so I think my services are not "active"?

Square brackets mean "explicitly active" ("marked" active). Implicitly active services will have curly brackets.

You should see at least one explicitly active service in the output of `dinitctl list`.

**Reply by iacore** [reply]
Why does some services stay in the output of `dinitctl list` after it is disabled? Example:

```
[     {-}] service name
```

It is different from an unloaded service, which is not shown in the list at all.

Is it possible to load service description from stdin? This will allow dinit to be used programmatically with dynamically generated services.

**Reply by davmac314** [reply]
> Why does some services stay in the output of dinitctl list after it is disabled?

Because "dinitctl list" lists the status of all loaded services

**Comment by iacore** [comment]
I regret Github Discussions already :(
